pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc

```diff
@@ -0,0 +1,621 @@
+
+    if (topicsNeedCreate->load() == 0) {
+        LOG_INFO("Successfully Subscribed to Topics");
+        setState(Ready);
+        if (!namespaceName_) {
+            namespaceName_ = TopicName::get(topic)->getNamespaceName();
+        }
+        multiTopicsConsumerCreatedPromise_.setValue(shared_from_this());
+        return;
+    }
+}
+
+// subscribe for passed in topic
+Future<Result, Consumer> MultiTopicsConsumerImpl::subscribeOneTopicAsync(const std::string& topic) {
+    TopicNamePtr topicName;
+    ConsumerSubResultPromisePtr topicPromise = boost::make_shared<Promise<Result, Consumer>>();
+    if (!(topicName = TopicName::get(topic))) {
+        LOG_ERROR("TopicName invalid: " << topic);
+        topicPromise->setFailed(ResultInvalidTopicName);
+        return topicPromise->getFuture();
+    }
+
+    if (state_ == Closed || state_ == Closing) {
+        LOG_ERROR("MultiTopicsConsumer already closed when subscribe.");
+        topicPromise->setFailed(ResultAlreadyClosed);
+        return topicPromise->getFuture();
+    }
+
+    // subscribe for each partition, when all partitions completed, complete promise
+    lookupServicePtr_->getPartitionMetadataAsync(topicName).addListener(
+        boost::bind(&MultiTopicsConsumerImpl::subscribeTopicPartitions, shared_from_this(), _1, _2, topicName,
+                    subscriptionName_, conf_, topicPromise));
+    return topicPromise->getFuture();
+}
+
+void MultiTopicsConsumerImpl::subscribeTopicPartitions(const Result result,
+                                                       const LookupDataResultPtr partitionMetadata,
+                                                       TopicNamePtr topicName,
+                                                       const std::string& consumerName,
+                                                       ConsumerConfiguration conf,
+                                                       ConsumerSubResultPromisePtr topicSubResultPromise) {
+    if (result != ResultOk) {
+        LOG_ERROR("Error Checking/Getting Partition Metadata while MultiTopics Subscribing- " << result)
+        topicSubResultPromise->setFailed(result);
+        return;
+    }
+
+    boost::shared_ptr<ConsumerImpl> consumer;
+    ConsumerConfiguration config;
+    ExecutorServicePtr internalListenerExecutor = client_->getPartitionListenerExecutorProvider()->get();
+
+    // all the consumers should have same name.
+    config.setConsumerName(conf_.getConsumerName());
+    config.setConsumerType(conf_.getConsumerType());
+    config.setBrokerConsumerStatsCacheTimeInMs(conf_.getBrokerConsumerStatsCacheTimeInMs());
+    config.setMessageListener(
+        boost::bind(&MultiTopicsConsumerImpl::messageReceived, shared_from_this(), _1, _2));
+    config.setReceiverQueueSize(conf_.getReceiverQueueSize());
+
+    int partitionsNumber = partitionMetadata->getPartitions() >= 1 ? partitionMetadata->getPartitions() : 1;
```

Comment:
rename to numPartitions

partitionsNumber implies it's a number assigned to the partitions rather than the number of partitions that exist.